src/analysis/modal_esm/predict.py

```diff
@@ -0,0 +1,222 @@
+                HuggingFace model identifier.
+
+        Notes:
+            - The `dtype` attribute is currently unused due to memory errors.
+        """
+        self.model_name = model_name
+
+    @modal.enter()
+    def load_model(self):
+        """Loads the ESM model and tokenizer into memory.
+
+        This method is called upon entering the Modal context. It loads the tokenizer and
+        model from the local cache, moves the model to the GPU, and sets it to evaluation mode.
+
+        Notes:
+            - The model and tokenizer are loaded from `MODEL_STORAGE_PATH`.
+            - Assumes that CUDA is available.
+        """
+        assert torch.cuda.is_available()
+
+        self.tokenizer = AutoTokenizer.from_pretrained(
+            self.model_name,
+            cache_dir=MODEL_STORAGE_PATH,
+            local_files_only=True,
+        )
+        self.model = EsmForMaskedLM.from_pretrained(
+            self.model_name,
+            cache_dir=MODEL_STORAGE_PATH,
+            local_files_only=True,
+        )
+        self.model.to("cuda")  # type: ignore
+        self.model.eval()
+
+    @modal.method()
+    def predict(
+        self,
+        sequence: str,
+        masked_positions: list[tuple[int, ...]],
+        max_length: int = 512,
+    ) -> np.ndarray:
+        """Calculates the logits for masked positions in a sequence.
+
+        Args:
+            sequence:
+                The input sequence as a string. Should be tokenizable by the model's tokenizer.
+            masked_positions:
+                A list of tuples, where each tuple contains positions (ints) in the
+                sequence to be masked. Each tuple corresponds to a separate observation
+                in the batch.
+            max_length:
+                The maximum sequence length. Sequences longer than this will be truncated.
+
+        Returns:
+            np.ndarray:
+                An array of logits from the model's output, with shape (batch_size,
+                sequence_length, vocab_size).
+        """
+        batch_size = len(masked_positions)
+
+        tokenized_sequence = self.tokenizer(
+            sequence,
+            return_tensors="pt",
+            padding=True,
+            truncation=True,
+            max_length=max_length,
+        )
+
+        input_ids = tokenized_sequence["input_ids"].to("cuda")
+        attention_mask = tokenized_sequence["attention_mask"].to("cuda")
+
+        input_ids = input_ids.repeat(batch_size, 1)
+        attention_mask = attention_mask.repeat(batch_size, 1)
+
+        # Apply mask tokens in batch for each pair of positions
+        for batch_idx, mask_positions in enumerate(masked_positions):
+            for pos in mask_positions:
+                input_ids[batch_idx, pos] = self.tokenizer.mask_token_id
+
+        with torch.no_grad():
+            outputs = self.model(input_ids=input_ids, attention_mask=attention_mask)
+            return outputs.logits.cpu().numpy()
```

Comment:
I'm pretty new to torch, so I'm not sure what best convention is (or what detach does). But after researching, it seems that detach is used to detach the tensor from the computation graph. But since this occurs within the `no_grad` context, there is no computation graph to detach from.

Reply:
Ah, right. TIL! (I somehow thought there was always a graph...)